src/test/java/io/jenkins/plugins/sshbuildagents/ssh/ConnectionImplTest.java

```diff
@@ -0,0 +1,205 @@
+                CredentialsScope.SYSTEM, "id", "", AgentConnectionBaseTest.USER, AgentConnectionBaseTest.PASSWORD);
+        connection.setCredentials(credentials);
+        int ret = connection.execCommand("echo FOO");
+        connection.close();
+        assertEquals(ret, 0);
+    }
+
+    @Test
+    public void testRunCommandSSHKey() throws IOException {
+        Connection connection = new ConnectionImpl(sshd.getHost(), sshd.getPort());
+        StandardUsernameCredentials credentials = new FakeSSHKeyCredential();
+        connection.setCredentials(credentials);
+        int ret = connection.execCommand("echo FOO");
+        connection.close();
+        assertEquals(ret, 0);
+    }
+
+    @Test
+    public void testCopyFile() throws IOException, FormException {
+        final File tempFile = tempFolder.newFile("tempFile.txt");
+        Connection connection = new ConnectionImpl(sshd.getHost(), sshd.getPort());
+        StandardUsernameCredentials credentials = new UsernamePasswordCredentialsImpl(
+                CredentialsScope.SYSTEM, "id", "", AgentConnectionBaseTest.USER, AgentConnectionBaseTest.PASSWORD);
+        connection.setCredentials(credentials);
+        String data = IOUtils.toString(getClass().getResourceAsStream("/fakeAgentJar.txt"), StandardCharsets.UTF_8);
+        connection.copyFile(tempFile.getAbsolutePath(), data.getBytes(StandardCharsets.UTF_8), true, true);
+        String dataUpload = FileUtils.readFileToString(tempFile, StandardCharsets.UTF_8);
+        assertEquals(data, dataUpload);
+    }
+
+    @Test
+    public void testShellChannel() throws IOException, FormException {
+        Logger logger = Logger.getLogger("io.jenkins.plugins.sshbuildagents.ssh.agents");
+        Connection connection = new ConnectionImpl(sshd.getHost(), sshd.getPort());
+        StandardUsernameCredentials credentials = new UsernamePasswordCredentialsImpl(
+                CredentialsScope.SYSTEM, "id", "", AgentConnectionBaseTest.USER, AgentConnectionBaseTest.PASSWORD);
+        connection.setCredentials(credentials);
+        ShellChannel shellChannel = connection.shellChannel();
+        shellChannel.execCommand("echo FOO");
+        byte[] data = IOUtils.readFully(
+                shellChannel.getInvertedStdout(),
+                shellChannel.getInvertedStdout().available());
+        String dataStr = IOUtils.toString(data, "UTF-8");
+        logger.info(dataStr);
+        assertEquals("FOO", StringUtils.chomp(dataStr));
+    }
+
+    @Test
+    public void testRunLongConnection() throws IOException, InterruptedException {
+        Connection connection = new ConnectionImpl(sshd.getHost(), sshd.getPort());
+        StandardUsernameCredentials credentials = new FakeSSHKeyCredential();
+        connection.setCredentials(credentials);
+        ShellChannel shellChannel = connection.shellChannel();
+        shellChannel.execCommand("sleep 500s");
+        for (int i = 0; i < 300; i++) {
```

Comment:
thanks that's plenty of time to have a coffee :)
do we really need this? Can't we skip this on normal dev life and run it on CI server only?